### PR TITLE
Rework executor

### PR DESCRIPTION
**Pull request by andll**
This pull requests simplifies narwhal executor - it changes both how executor works internally and how narwhal executor interacts with consumer.

### Changes to sui interface:
* Pass transaction as `Vec<u8>`. This pushes serialization-deserialization fully to consumer. Previously serialization was done on consumer and de-serialization was done in narwhal, which does not make sense really
* Remove _executor->consumer_ channel. Currently both channel and callback existed as part of narwhal executor interface, this is redundancy that is not needed and callback can support execution channel if needed.
* Do not return `Outcome` from callback - this is not used once channel above is removed
* Do not return error from `load_execution_indices` - there is nothing to do with this error on nw side except to panic, so we can as well panic inside sui handler
* Do not return error from `handle_consensus_transaction`- entirety of error handling can be done in callback.
* Remove `ask|release_consensus_write_lock`  - this can be simply replaced by rust owner semantics

### How new executor works:
New executor only have 2 simple tasks.
* **Subscriber** task:
   * Download payload in memory(see below)
      * Buffer up to N downloads in parallel via FuturesOrdered
      * Send downloaded batch to notifier task
      * Propagate back-pressure to consensus if more then N downloads are pending
* **Notifier** task:
   * For each payload, execute consumer callback
   * Propagate back pressure from callback to subscriber

### Payload download 
is done in following manner:
* Local worker is checked first, in case payload is present locally
* If not found in local worker, start issuing parallel requests to 2f+1 remote workers signed the certificate
* Each request starts with staggering delay and performs infinite retries with exponential backoff
* Staggering delay allows not to have too many requests in case of good network conditions
* Exponential backoff allows not to spam too many request - after ramp up we don't issue more then 1 request per minute to remote worker (per download, so multiply by N max pending downloads for total)
* When one of requests succeeds all pending requests are cancelled

To recap, payload download is done as following formula:

```
Fetch: Select[exponential_backoff_with_stagger(get_payload_rpc()) for each authority signed certificate]
```

### Restartability
New executor only uses certificate index in order to support restart. Consumer must be idempotent in handling repeated transactions if execution was aborted in the middle of the certificate/payload.

### Main differences
between new and old executor

* New executor does not issue availability network requests to remote authorities, leveraging certificate signatures instead
* No temporary storage, we only hold up to N batches in memory after download. This reduces IO and complexity
* Not using batch_(loader|synchronizer) and uses simple network requests instead. No nested network calls and nested timeouts.

### Bounded resource usage
* Total number of pending downloads is limited by `MAX_PENDING_PAYLOADS`
* Total number of batches in memory is limited by `subscriber->notifier channel size` + `MAX_PENDING_PAYLOADS`
* Total number of pending RPC requests from this component: `MAX_PENDING_PAYLOADS`*`(2f+1)`
* Total number of pending RPC requests to a given node: `MAX_PENDING_PAYLOADS`

### TODO
 - [x] Rename GetPayload -> BatchRequest
 - [x] Rollback log level change
 - [x] Tests
 - [x] Randomize authorities
 - [x] Verify payload match digest
 - [x] Metrics
 - [x] Tracing

### CODE REVIEW

File: node/src/lib.rs
Comment (mwtian) [inline node/src/lib.rs:9]: Are we able to delete `temp_batch_store`, `temp_batch_map` and `TEMP_BATCH_CF` from this file?

File: executor/src/subscriber.rs
Comment (mwtian) [inline executor/src/subscriber.rs:217]: I think this was the only usage of `BlockCommand::GetBlock`. We can add a comment above `BlockCommand::GetBlock` to mention that it is now unused and see if we can delete unnecessary code.

**Review by mwtian — APPROVED** [review]
Great cleanup!

File: executor/src/lib.rs
Comment (bmwill) [inline executor/src/lib.rs:65]: Why isn't the network able to be directly passed in and instead is a oneshot?

Comment (bmwill) [reply executor/src/lib.rs:65]: Ah, ok I see. The executor is started outside the primary and the network is constructed inside the primary which makes the dependency difficult to have at the time of construction.

File: executor/src/subscriber.rs
Comment (mwtian) [inline executor/src/subscriber.rs:209]: Do we need to randomize the order of authorities to avoid overloading the authority that appears first? Also the order of authority to query may be relevant depending on how we handle response from byzantine workers.

File: executor/src/subscriber.rs
Comment (mwtian) [inline executor/src/subscriber.rs:273]: Should we verify the payload actually corresponds to the digest here? I'm not sure if this verification is done at a higher or lower level.

File: executor/src/subscriber.rs
Comment (mwtian) [inline executor/src/subscriber.rs:274]: Maybe we don't need to keep retrying on a known byzantine worker, and just return and add a loop to the higher level `select_all()`.

**Review by mwtian — COMMENTED** [review]
Btw, if we continue on this strategy, maybe we can optimize / simplify [this](https://github.com/mwtian/narwhal/blob/40f33bdc251b1e81b0194a559fa0a54ad1bc64dc/primary/src/header_waiter.rs#L177-L218) as a follow up.

**Comment by andll** [comment]
Updated - added tests, randomizing authorities, verifying response digest

File: executor/src/subscriber.rs
Comment (andll) [reply executor/src/subscriber.rs]: I think it is ok to try again - maybe there was some deployment issue on the validator and it got fixed. We have exponential backoff in any case, so it won't be too many retries

File: executor/src/subscriber.rs
Comment (andll) [reply executor/src/subscriber.rs:217]: I will take a look if block waiter can be removed after this PR (also will need to sync with @akichidis about it)

File: node/src/lib.rs
Comment (andll) [reply node/src/lib.rs:9]: I'll take a look at that, I think right now there is still some code in block synchronizer that might use it(but yes, we should delete it)

File: executor/src/lib.rs
Comment (akichidis) [inline executor/src/lib.rs:156]: @andll out of curiosity why is this necessary? Am I getting it right that is mainly for convenience reasons or am I missing something?

File: executor/src/state.rs
Comment (akichidis) [inline executor/src/state.rs:9]: Since you removed the tests I would recommend to remove the implementation methods of `ExecutionIndices` as well since they don't seem to be used anymore.

**Review by akichidis — COMMENTED** [review]
Great work @andll !! Left some comments. I'll wait for the rest of the tests your are working on - but that looks good!

File: executor/src/subscriber.rs
Comment (akichidis) [reply executor/src/subscriber.rs:219]: I like the staggering though, great approach!

File: executor/src/lib.rs
Comment (andll) [reply executor/src/lib.rs:156]: Yeah, idea is that some places will use Arc<X> as executor state, and some(perhaps unit tests) might want to pass type without Arc

**Comment by andll** [comment]
Re 'why warn vs debug' - there is separate commit that changes bunch of debug to warn, i'll revert it before merging once we have debug logs enabled for executor in private testnet

**Comment by andll** [comment]
@akichidis I think I addressed rest of your comments

File: executor/src/subscriber.rs
Comment (asonnino) [inline executor/src/subscriber.rs:268]: If this happens it seems we have a serious problem, don't we? Should we log a warn / error?

**Comment by akichidis** [comment]
> @akichidis I think I addressed rest of your comments, but let me know if something is missing

Great @andll, reviewing now

File: executor/src/subscriber.rs
Comment (akichidis) [inline executor/src/subscriber.rs:57]: @andll are you going to add any test our the subscriber it self? I think it would be great if we could have some unit test coverage for that as well to ensure we don't accidentally break the behaviour

File: executor/src/subscriber.rs
Comment (akichidis) [reply executor/src/subscriber.rs:268]: Well I guess not necessarily - workers of other validators might not respond in time (for whatever reasons)

File: executor/src/subscriber.rs
Comment (akichidis) [inline executor/src/subscriber.rs:228]: Out of curiosity, will this eventually timeout at some point if no response is received? Not sure how Anemo handles requests (maybe @bmwill can provide some pointers?)

**Review by akichidis — APPROVED** [review]
Thanks for the changes @andll ! This PR looks great 💯  

I am approving to make sure is not getting blocked, but I would highly suggest to increase the test coverage in `Subscriber` & the `Notifier`. Right now there are methods & scenarios (ex retries) which are not covered and I would feel safer if we had something in place (especially to ensure that we don't break things as the codebase evolves)

File: executor/src/subscriber.rs
Comment (bmwill) [reply executor/src/subscriber.rs:228]: If the connection is broken then yes it'll eventually timeout, otherwise if you need a particular timeout for this rpc you'll want to wrap it in tokio::time::timeout().

**Comment by andll** [comment]
@akichidis @asonnino - I'll merge this PR to unblock nw<->sui merge. Will reply to your comments and/or make follow up PR after that

File: executor/src/subscriber.rs
Comment (andll) [reply executor/src/subscriber.rs:268]: I don't have strong opinion on log level here - do you know how we typically log timeouts in other components?

File: executor/src/subscriber.rs
Comment (andll) [reply executor/src/subscriber.rs:219]: So in ideal scenario we want to issue one RPC, and if node is healthy get result and move on (and if staggering step is chosen well this is what will happen with current approach)
In less ideal scenario we do want to have some overlapping attempts, because waiting exponential backoff before trying next node will be sub-optimal.

I think overall what you suggesting is very similar to how staggering delay and exponential back off actually end up working (except that we will have some overlapping requests).

**Pull request merged**
